### DISCUSSION TITLE
Question about using plugin's internal functions

### DISCUSSION THREAD

**Discussion by augustocdias** [williamboman/nvim-lsp-installer#856] [Q&A] [answered]
Hi,

I'm trying to use some internal functions from the plugin to complement a LSP server and I'm not sure how to deal with the futures/promises that all functions return. 

Basically for rust-tools there's a possibility to improve the debugging experience by providing the location of a vscode extension installation to the LSP client. I was willing to automate this for:

- Download the extension
- Uncompress the required files
- Put in a specific folder

Instead of implementing everything by hand I thought of using functions of lsp installer but I'm stuck on how to deal with the futures.

Basically I start by checking if there are updates. I basically create a folder with the version as the name. I the `fetch_latest_release` to get that info. If there's a new version, I delete the existing folder with `spawn.rm`.  I download the asset with `fetch` and finally I call `spawn.unzip` to decompress the files I need in the correct directory. 

All this calls from what I could see are `async`. How do I await on them? One depends on the other, so they need to be sequential. Also I would like to not block neovim with this calls if possible. Right now I just call them but I get nothing because I think the futures are not being executed. 

Any help would be appreciated. 

Here's my current code... I only get the prints of "install folder created" and "finished".

```lua
local repo = 'vadimcn/vscode-lldb'
local github_client = require('nvim-lsp-installer.core.managers.github.client')
local spawn = require('nvim-lsp-installer.core.spawn')
local async = require('nvim-lsp-installer.core.async')
local fetch = require('nvim-lsp-installer.core.fetch')

local install_path = vim.fn.stdpath('data') .. '/vscode-lldb/'

local function exists(file)
    local ok, err, code = os.rename(file, file)
    if not ok then
        if code == 13 then
            -- Permission denied, but it exists
            return true
        end
    end
    return ok, err
end

local function check_download_new_release()
    local latest = github_client.fetch_latest_release(repo)
    local tag_name = latest.tag_name
    print('tag name ' .. tag_name)
    if not exists(install_path .. tag_name) then
        spawn.rm({ '-rf ', install_path .. '*' })
        spawn.mkdir({ install_path .. tag_name })
        print('directories created')
        local download_url = ('https://github.com/%s/releases/download/%s/%s'):format(
            repo,
            tag_name,
            'codelldb-aarch64-darwin.vsix'
        )
        fetch(download_url, {
            out_file = install_path .. 'codelldb.zip',
        })
            :map_err(function(err)
                return ('Failed to download file %q.\n%s'):format(download_url, err)
            end)
            :get_or_throw()
        return install_path .. tag_name, install_path .. 'codelldb.zip'
    else
        return false
    end
end

local function create_install_dir()
    if not exists(install_path) then
        spawn.mkdir({ install_path })
    end
end

local function install(file, target_dir)
    spawn.unzip({ '-j', file, 'extension/adapter/codelldb', 'extension/lldb/lib/liblldb.dylib', '-d', target_dir })
    spawn.rm({ '-rf', file })
end

return {
    install_codelldb = function()
        async.run(function()
            create_install_dir()
            print('install folder created')
            local new_release_dir, zip_file = check_download_new_release()
            if new_release_dir then
                print('installing')
                install(zip_file, new_release_dir)
            end
        end, function()
            print('finished')
        end)
    end,
}
```

**Comment by williamboman** [comment, marked as answer]
Hello! I just happened to make https://github.com/williamboman/mason.nvim public the other day. It provides `codelldb` as a package available for installation, might be of interest.

> Here's my current code... I only get the prints of "install folder created" and "finished".

The second argument to `async.run()` is a callback that'll be invoked with 2 arguments pcall-style `(success, value)`. Most likely there's an error occurring somewhere, which won't bubble up to the main event loop. I'd actually recommend using the `async.scope` function instead if you don't need to access the return value of the async function - it will ensure to log errors for you:

```lua
return {
    install_codelldb = async.scope(function()
          create_install_dir()
          print('install folder created')
          local new_release_dir, zip_file = check_download_new_release()
          if new_release_dir then
              print('installing')
              install(zip_file, new_release_dir)
          end)
    end),
}
```

**Reply by augustocdias** [reply]
Is Mason a replacement for LspInstaller? They seem to overlap in features. And the codelldb you donwload is not the same I need. You download from Microsoft. And I don't need the whole package, I basically need 2 binaries from it. 

I'll try your suggestion with the scope.

**Reply by williamboman** [reply]
> Is Mason a replacement for LspInstaller?

Yeah it supports all servers in nvim-lsp-installer. It's still pretty early stage so I'm just looking to recruit testers and gather early feedback right now. I don't intend on actively working on nvim-lsp-installer going forward

> And the codelldb you donwload is not the same I need. You download from Microsoft. 

It downloads the release files from https://github.com/vadimcn/vscode-lldb, I just noticed the description and homepage [were incorrect](https://github.com/williamboman/mason.nvim/pull/25/files) 🙈

**Reply by augustocdias** [reply]
Cool. I'll take a look on Mason later. One thing I'm missing is documentation on how to use the downloaded servers... With lsp-install we basically "tie" it to lsp config together...

One last question. To return a value from my function would be just using the callback?

**Reply by augustocdias** [reply]
Looking into Mason's code I see a lot of interesting stuff but no idea how they're used or how to use them. For example the schemas for the configuration of LSPs. Is there any integration with LSP-Config like LSP-Installer has or it just provides the executables in the path? In case of vscode-lldb, rust-tools needs the location of binary and of one lib. How do I locate them? Is there any API that would give me their locations back?

**Reply by williamboman** [reply]
> For example the schemas for the configuration of LSPs

The schemas are only used for a read-only, visual, representation of the server's setting. It's used in lsp-installer as well as Mason to render a visual, semi-interactive, representation in the UI window.

> Is there any integration with LSP-Config like LSP-Installer has or it just provides the executables in the path?

It does, via the `mason-lspconfig ` extension (which mimics what exists in lsp-installer). It's worth noting that this only really applies to a few servers that require extra configuration to be functional - the majority of lsp-installer's "integration" with lspconfig is to modify the PATH used by lspconfig, this is not needed with mason because it links everything to a global PATH. I'll need to write some more docs though.

> In case of vscode-lldb, rust-tools needs the location of binary and of one lib. How do I locate them? Is there any API that would give me their locations back?

The binary will already be in PATH so most likely the canonical name `codelldb` should suffice. You could do something like this: https://github.com/williamboman/mason.nvim/discussions/33. I'm also looking into whether it'd make sense to introduce a `mason-dap` extension that'd help bridge these things with other popular 3rd party plugins.

**Reply by augustocdias** [reply]
> It does, via the mason-lspconfig extension (which mimics what exists in lsp-installer). It's worth noting that this only really applies to a few servers that require extra configuration to be functional - the majority of lsp-installer's "integration" with lspconfig is to modify the PATH used by lspconfig, this is not needed with mason because it links everything to a global PATH. I'll need to write some more docs though.



I know
But I do some customization nevertheless. For example I check server capabilities before setting some auto commands. Also for rust I don't use lsp config directly but rather rust tools. How would I integrate with it?

**Reply by williamboman** [reply]
You set it up as you’d normally do, nothing special needed.

**Reply by augustocdias** [reply]
I'm finally moving from lsp-installer to mason and I have a question. I saw you deprecated the old way of having a callback to setup lspconfig in favor of setting up a hook in lspconfig for mason (and previously lsp-installer) to install the lsp server.

How does that work with plugins like rust-tools and jdtls that wraps lspconfig? If I setup up them will it work the same way? Mason will detect and install the required binaries automatically?
And for non LSP stuff like null-ls and the codelldb as I was asking previously? I should manually install or is there a way to automate the installation through a script?

**Reply by williamboman** [reply]
There's `:h mason-lspconfig.setup_handlers()`. As long as wrapper plugins call `lspconfig[server_name].setup()` it should work just the same, where Mason will be able to detect these servers (as long as you set up mason before you set up any servers of course).

> And for non LSP stuff like null-ls and the codelldb as I was asking previously? I should manually install or is there a way to automate the installation through a script?

As of right now, manually. Mason comes with a Lua API which you can use to easily code up a custom snippet that installs everything you'd like
